testnet/stacks-node/src/tests/nakamoto_integrations.rs

```diff
@@ -5003,3 +5003,277 @@ fn signer_chainstate() {
+        &[sender_signer_sk],
+        &signers,
+    );
+
+    info!("Pausing commit ops to trigger a tenure extend.");
+    TEST_SKIP_COMMIT_OP.lock().unwrap().replace(true);
+
+    next_block_and(&mut btc_regtest_controller, 60, || Ok(true)).unwrap();
+
+    signer_vote_if_needed(
+        &btc_regtest_controller,
+        &naka_conf,
+        &[sender_signer_sk],
+        &signers,
+    );
+
+    // Submit a TX
+    let transfer_tx = make_stacks_transfer(&sender_sk, 0, send_fee, &recipient, send_amt);
+    let transfer_tx_hex = format!("0x{}", to_hex(&transfer_tx));
+
+    let tip = NakamotoChainState::get_canonical_block_header(chainstate.db(), &sortdb)
+        .unwrap()
+        .unwrap();
+
+    let mut mempool = naka_conf
+        .connect_mempool_db()
+        .expect("Database failure opening mempool");
+
+    mempool
+        .submit_raw(
+            &mut chainstate,
+            &sortdb,
+            &tip.consensus_hash,
+            &tip.anchored_header.block_hash(),
+            transfer_tx.clone(),
+            &ExecutionCost::max_value(),
+            &StacksEpochId::Epoch30,
+        )
+        .unwrap();
+
+    next_block_and_process_new_stacks_block(&mut btc_regtest_controller, 60, &coord_channel)
+        .unwrap();
+
+    signer_vote_if_needed(
+        &btc_regtest_controller,
+        &naka_conf,
+        &[sender_signer_sk],
+        &signers,
+    );
+
+    next_block_and(&mut btc_regtest_controller, 60, || Ok(true)).unwrap();
+
+    signer_vote_if_needed(
+        &btc_regtest_controller,
+        &naka_conf,
+        &[sender_signer_sk],
+        &signers,
+    );
+
+    info!("Resuming commit ops to mine regular tenures.");
+    TEST_SKIP_COMMIT_OP.lock().unwrap().replace(false);
+
+    // Mine 15 more regular nakamoto tenures
+    for _i in 0..15 {
+        let commits_before = commits_submitted.load(Ordering::SeqCst);
+        let blocks_processed_before = coord_channel
+            .lock()
+            .expect("Mutex poisoned")
+            .get_stacks_blocks_processed();
+        next_block_and(&mut btc_regtest_controller, 60, || {
+            let commits_count = commits_submitted.load(Ordering::SeqCst);
+            let blocks_processed = coord_channel
+                .lock()
+                .expect("Mutex poisoned")
+                .get_stacks_blocks_processed();
+            Ok(commits_count > commits_before && blocks_processed > blocks_processed_before)
+        })
+        .unwrap();
+
+        signer_vote_if_needed(
+            &btc_regtest_controller,
+            &naka_conf,
+            &[sender_signer_sk],
+            &signers,
+        );
+    }
+
+    // load the chain tip, and assert that it is a nakamoto block and at least 30 blocks have advanced in epoch 3
+    let tip = NakamotoChainState::get_canonical_block_header(chainstate.db(), &sortdb)
+        .unwrap()
+        .unwrap();
+
+    // assert that the transfer tx was observed
+    let transfer_tx_included = test_observer::get_blocks()
```

Comment:
I'm not sure this is sufficient for testing. You'd want to confirm that a tenure-extend got issued for the empty sortition.

Also, I think you'll want the following additional tests:

* There are two miners, and every other snapshot is empty.  The active miner issues tenure-extends whenever it sees an empty sortition.

* The miner issues a BlockFound tenure change transaction for each sortition it wins, even if those sortitions are consecutive.

Reply:
The updated integration test:

1. asserts that a tenure-extend was issued.
2. asserts that the miner issues block-found tenure change transactions for the other sortitions.

As for testing a scenario with multiple miners -- I suppose that could be possible with the integration testing framework, but its not readily achieved with the current setup. The multiple miner scenario isn't really unique to this PR: even for "block found" tenures, the relayer has to check if it won the block, so the same "am I the correct miner?" branch coverage is missing there. My suggestion is that we create an issue for "Add nakamoto integration tests for multi-miner scenario" and separate it from this PR.

Reply:
I am fine wiht this because I am struggling hard and it would be nice to not have this PR blocked. See https://github.com/stacks-network/stacks-core/issues/4896